File: backend/src/database/repos/memory.ts
import { ConflictError } from "../../common/errors.js";
import type { Employee, ExchangeRate, ListFilters, SalaryRecord } from "../../common/types.js";
import { currentSalary } from "../../validators/salaries.js";
import type { EmployeeRepository } from "./employees.js";
import type { FxRepository } from "./fx.js";
import type { SalaryRepository } from "./salaries.js";

export class MemoryEmployeeRepository implements EmployeeRepository {
  constructor(private readonly rows: Employee[] = []) {}

  async get(id: string) {
    return this.rows.find((row) => row.id === id) ?? null;
  }

  private byCode(code: string) {
    return this.rows.find((row) => row.employeeCode === code);
  }

  private byEmail(email: string) {
    const needle = email.toLowerCase();
    return this.rows.find((row) => row.email.toLowerCase() === needle);
  }

  async add(employee: Employee) {
    const last = [...this.rows]
      .map((row) => row.employeeCode)
      .filter((code) => code.startsWith("ACME-"))
      .sort()
      .at(-1);
    const next =
      employee.employeeCode ||
      `ACME-${String((last ? Number(last.split("-")[1]) : 0) + 1).padStart(5, "0")}`;
    if (this.byEmail(employee.email)) {
      throw new ConflictError("email already exists");
    }
    if (this.byCode(next)) {
      throw new ConflictError(`employee_code ${next} already exists`);
    }
    const stored = { ...employee, employeeCode: next };
    this.rows.push(stored);
    return stored;
  }

  async save(employee: Employee) {
    const other = this.byEmail(employee.email);
    if (other && other.id !== employee.id) {
      throw new ConflictError("email already exists");
    }
    const index = this.rows.findIndex((row) => row.id === employee.id);
    if (index >= 0) {
      this.rows[index] = employee;
    }
    return employee;
  }

  async listPage(filters: ListFilters) {
    let items = [...this.rows];
    if (filters.names.length > 0) {
      items = items.filter((row) =>
        filters.names.some((name) => row.name.toLowerCase().includes(name.toLowerCase())),
      );
    }
    if (filters.countries.length > 0) {
      items = items.filter((row) => filters.countries.includes(row.countryCode));
    }
    if (filters.departments.length > 0) {
      items = items.filter((row) => filters.departments.includes(row.department));
    }
    if (filters.statuses.length > 0) {
      items = items.filter((row) => filters.statuses.includes(row.status));
    }
    items.sort((a, b) => a.name.localeCompare(b.name));
    return {
      items: items.slice(filters.offset, filters.offset + filters.limit),
      total: items.length,
    };
  }

  async listActive() {
    return this.rows.filter((row) => row.status === "active");
  }
}

export class MemorySalaryRepository implements SalaryRepository {
  constructor(private readonly rows: SalaryRecord[] = []) {}

  async add(record: SalaryRecord) {
    if (
      this.rows.some(
        (row) => row.employeeId === record.employeeId && row.effectiveFrom === record.effectiveFrom,
      )
    ) {
      throw new ConflictError(
        "an employee cannot have two salary records with the same effective_from",
      );
    }
    this.rows.push(record);
    return record;
  }

  async listForEmployee(employeeId: string) {
    return this.rows
      .filter((row) => row.employeeId === employeeId)
      .sort((a, b) => b.effectiveFrom.localeCompare(a.effectiveFrom));
  }

  async currentForMany(employeeIds: string[], today: string) {
    const latest = new Map<string, SalaryRecord>();
    for (const id of employeeIds) {
      const current = currentSalary(
        this.rows.filter((row) => row.employeeId === id),
        today,
      );
      if (current) {
        latest.set(id, current);
      }
    }
    return latest;
  }
}

export class MemoryFxRepository implements FxRepository {
  constructor(private readonly rows: ExchangeRate[] = []) {}

  async upsert(rate: ExchangeRate) {
    const index = this.rows.findIndex((row) => row.currency === rate.currency);
    if (index >= 0) {
      this.rows[index] = rate;
    } else {
      this.rows.push(rate);
    }
  }

  async list() {
    return [...this.rows];
  }
}

File: backend-py/app/repos/memory.py
from __future__ import annotations

from app.errors import ConflictError
from app.models import Employee, ExchangeRate, ListFilters, SalaryRecord
from app.validators import pick_current_salary


class MemoryEmployeeRepository:
    def __init__(self, rows: list[Employee] | None = None):
        self.rows = rows or []

    def get_by_id(self, employee_id: str) -> Employee | None:
        return next((row for row in self.rows if row.id == employee_id), None)

    def _find_by_code(self, code: str) -> Employee | None:
        return next((row for row in self.rows if row.employee_code == code), None)

    def _find_by_email(self, email: str) -> Employee | None:
        needle = email.lower()
        return next((row for row in self.rows if row.email.lower() == needle), None)

    def insert(self, employee: Employee) -> Employee:
        last = sorted(
            (row.employee_code for row in self.rows if row.employee_code.startswith("ACME-"))
        )
        last_code = last[-1] if last else None
        suffix = int(last_code.split("-")[1]) if last_code else 0
        nxt = employee.employee_code or f"ACME-{suffix + 1:05d}"
        if self._find_by_email(employee.email):
            raise ConflictError("email already exists")
        if self._find_by_code(nxt):
            raise ConflictError(f"employee_code {nxt} already exists")
        stored = Employee(
            id=employee.id,
            employee_code=nxt,
            name=employee.name,
            email=employee.email,
            country_code=employee.country_code,
            department=employee.department,
            designation=employee.designation,
            status=employee.status,
            created_at=employee.created_at,
            updated_at=employee.updated_at,
        )
        self.rows.append(stored)
        return stored

    def update(self, employee: Employee) -> Employee:
        other = self._find_by_email(employee.email)
        if other and other.id != employee.id:
            raise ConflictError("email already exists")
        for i, row in enumerate(self.rows):
            if row.id == employee.id:
                self.rows[i] = employee
                break
        return employee

    def list_page(self, filters: ListFilters) -> tuple[list[Employee], int]:
        items = list(self.rows)
        if filters.names:
            items = [
                row
                for row in items
                if any(name.lower() in row.name.lower() for name in filters.names)
            ]
        if filters.countries:
            items = [row for row in items if row.country_code in filters.countries]
        if filters.departments:
            items = [row for row in items if row.department in filters.departments]
        if filters.statuses:
            items = [row for row in items if row.status in filters.statuses]
        items.sort(key=lambda row: row.name)
        return items[filters.offset : filters.offset + filters.limit], len(items)

    def list_active(self) -> list[Employee]:
        return [row for row in self.rows if row.status == "active"]


class MemorySalaryRepository:
    def __init__(self, rows: list[SalaryRecord] | None = None):
        self.rows = rows or []

    def insert(self, record: SalaryRecord) -> SalaryRecord:
        if any(
            row.employee_id == record.employee_id and row.effective_from == record.effective_from
            for row in self.rows
        ):
            raise ConflictError(
                "an employee cannot have two salary records with the same effective_from"
            )
        self.rows.append(record)
        return record

    def list_for_employee(self, employee_id: str) -> list[SalaryRecord]:
        return sorted(
            [row for row in self.rows if row.employee_id == employee_id],
            key=lambda row: row.effective_from,
            reverse=True,
        )

    def get_current_for_many(self, employee_ids: list[str], today: str) -> dict[str, SalaryRecord]:
        latest: dict[str, SalaryRecord] = {}
        for employee_id in employee_ids:
            current = pick_current_salary(
                [row for row in self.rows if row.employee_id == employee_id],
                today,
            )
            if current:
                latest[employee_id] = current
        return latest


class MemoryFxRepository:
    def __init__(self, rows: list[ExchangeRate] | None = None):
        self.rows = rows or []

    def upsert(self, rate: ExchangeRate) -> None:
        for i, row in enumerate(self.rows):
            if row.currency == rate.currency:
                self.rows[i] = rate
                return
        self.rows.append(rate)

    def list(self) -> list[ExchangeRate]:
        return list(self.rows)
